programs/cleos/do_http_post_libcurl.cpp

```diff
@@ -0,0 +1,142 @@
+            fputc(x, stream);
+         }
+
+         fputc('\n', stream); /* newline */
+      }
+   }
+
+   int my_trace(CURL* handle, curl_infotype type, char* data, size_t size, void* userp) {
+      const char* text;
+      (void)handle; /* prevent compiler warning */
+      (void)userp;
+
+      switch (type) {
+         case CURLINFO_TEXT: fprintf(stderr, "== Info: %s", data);
+         default: /* in case a new one is introduced to shock us */ return 0;
+
+         case CURLINFO_HEADER_OUT: text = "=> Send header"; break;
+         case CURLINFO_DATA_OUT: text = "=> Send data"; break;
+         case CURLINFO_SSL_DATA_OUT: text = "=> Send SSL data"; break;
+         case CURLINFO_HEADER_IN: text = "<= Recv header"; break;
+         case CURLINFO_DATA_IN: text = "<= Recv data"; break;
+         case CURLINFO_SSL_DATA_IN: text = "<= Recv SSL data"; break;
+      }
+
+      dump(text, stderr, (unsigned char*)data, size);
+      return 0;
+   }
+
+   struct curl_handle : std::unique_ptr<CURL, decltype(&curl_easy_cleanup)> {
+      curl_handle(int) : std::unique_ptr<CURL, decltype(&curl_easy_cleanup)>( curl_easy_init(), &curl_easy_cleanup) {}
+   };
+
+
+   std::tuple<unsigned int, std::string> do_http_post(const std::string& base_uri, const std::string& path,
+                                                      const std::vector<std::string>& headers,
+                                                      const std::string& postjson, bool verify_cert, bool verbose,
+                                                      bool trace) {
+
+      static bool initialized = false;
+
+      if (!initialized) {
+         auto res = curl_global_init(CURL_GLOBAL_DEFAULT);
+         EOS_ASSERT(res == CURLE_OK, chain::http_exception, curl_easy_strerror(res));
+         initialized = true;
+      }
+
+      static std::map<std::string, curl_handle> handles;
+      auto [itr, _] = handles.emplace(base_uri, 0);
+      auto curl = itr->second.get();
+      EOS_ASSERT(curl != 0, chain::http_exception, "curl_easy_init failed");
+
+      std::string uri;
+
+      const char* unix_socket_prefix     = "unix://";
+      const int   unix_socket_prefix_len = strlen(unix_socket_prefix);
+
+      if (strncmp(unix_socket_prefix, base_uri.c_str(), unix_socket_prefix_len) == 0) {
+         curl_easy_setopt(curl, CURLOPT_UNIX_SOCKET_PATH, base_uri.c_str() + unix_socket_prefix_len);
+         uri = "http://localhost" + path;
+      } else {
+         uri = base_uri + path;
+      }
+
+      curl_easy_setopt(curl, CURLOPT_URL, uri.c_str());
+      curl_easy_setopt(curl, CURLOPT_POST, 1L);
+      curl_easy_setopt(curl, CURLOPT_HTTP_VERSION, (long)CURL_HTTP_VERSION_1_1);
+      curl_easy_setopt(curl, CURLOPT_POSTFIELDSIZE, postjson.size());
+      curl_easy_setopt(curl, CURLOPT_POSTFIELDS, postjson.c_str());
+
+      if (trace) {
+         curl_easy_setopt(curl, CURLOPT_DEBUGFUNCTION, my_trace);
+         curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 1L);
```

Comment:
Is there a reason `CURLOPT_FOLLOWLOCATION` is (only) enabled in trace mode? Such an option changes the semantics of the request -- it's not strictly a debug option.